### ISSUE TITLE
Proposal: Rename the raw_t codec and change its return type (and alter the functionality a bit ...)

### ISSUE THREAD

**Issue by punchfox** [spotify/spotify-json#16]
I would like to explore some changes to the `raw_t` codec type before we settle on version 1.0. I feel that the name is almost an implementation detail rather than describing what it really is. Yes, it allows you to access the raw bytes of the value as encoded in the file, but what it is really useful for is to represent any value that can later be decode by someone else.

I think it would be better named something like `value_t`, `any_value_t`, `anything_t` etc., returning a `encoded_value<T>`, where `T` is the storage type, e.g., `ref`, `std::string`, `std::vector`. The `encoded_value` will have an implicit casting operator to `const &T` (see the section about `json::encoded` for why this is useful).

**Question:** Would we allow moving a `T` out of the `encoded_value<T>`?

Using a special type (`encoded_value`) instead of a raw `std::string` or `std::vector` allows us to build in validation that the JSON we are writing is in fact correct. The current `raw_t` codec will happily trust that the caller is passing valid JSON, something which we already have found led to a bug (not deployed). Anything that is already an `encoded_value` is known to be valid JSON, so validation is free. When creating an `encoded_value` from a custom `std::string` or `std::vector` we'll validate immediately in the constructor.

I also propose that we change the return type of `json::encode()` to `encoded_value<std::string>` (instead of `std::string`). With the implicit cast to `std::string`, using it should feel almost the same as today, assigning directly into an `std::string`. It would be more expensive though, since we might not be able to move the string out of the value type (see question above).

The `encoded_value` type might have a `type` property to identify what kind of JSON value it stores. This is nice to have but optional. The type of the `type` property would be something like:

``` c++
enum class type {
  null,
  boolean,
  number,
  string,
  object,
  array
};
```

With all these changes (not considering the `type` property), it would be easier to store and work with values that the decoder does not know the type of. A _Spotify-internal_ example of this would be track metadata, which we now could store like this (here `encoded_value` has a default storage type):

``` c++
std::unordered_map<std::string, json::encoded_value<>> metadata = json::decode(...);
...
const auto duration = json::decode<std::chrono::milliseconds>(metadata["duration"]);
const auto title = json::decode<std::string>(metadata["title"]);
...
metadata["explicit"] = json::encode(true);
```

**Question:** In the example above, `encoded_value` is default constructible. That what? `"null"`?

Example of JSON validation:

``` c++
encoded_value<>("[1, 2, 3]") == "[1, 2, 3]";
encoded_value<>("[1, 2, 3");  // throws an exception
```

What do you think about this? Many details to work out. I do think it is a nicer and more powerful system than what we have in place today.

**Comment by emil-e** [comment]
I like the concept, I say go. However, for this to be useful, it has to be waterproof, not water resistant... so to speak.

**Regarding question 1:**
We can move it. We can have an overloaded conversion operator of the form:
`operator T &&() &&`

This feature of C++ is called "ref qualifiers" and it lets us have this overload only be picked in case of rvalues. These occur in two places, the first is "natural rvalues":
```
store_somewhere(json::encode(my_huge_value));
```

In this case, `json::encode` returns a `encoded_value<std::string>` by value which in this context is an rvalue. Thus, the `T &&() &&` overload is picked and the returned `std::string` is _also_ an rvalue which is swiftly moved into the call of `store_somewhere`. After this move, the `encoded_value` is no longer valid which is fine since it's an rvalue which per definition will be destroyed afterwards.

The second case is, of course, the case of creating an rvalue through `std::move`:
```
auto encoded = json::encode(my_huge_value);
store_somewhere(std::move(encoded));
```

The fact that we're using `std::move` instead doesn't really change anything. After using `std::move`, one should not assume anything about the value so it being invalid after the move is not a problem, I would say. If one wants to be a semantics lawyer, there has been a debate about whether one should be able to assume anything about move-from types but I think this wouldn't really surprise anyone who is reasonably proficient in `move`ing stuff around.

**Regarding question 2:**
Does it have to be default constructible? I guess it is more convenient if it is. In alignment with my statement above about waterproofing, it would have to be valid JSON, yes. But `{}` could also be reasonable and it is shorter. Isn't it also the case that an empty object is valid JSON but `null` standalone is not?


**Further question 1:**
When decoding, it would be typical to get an `encoded_value<ref>` in order to avoid copying. However, when _encoding_, you're not likely to get an `encoded_value<ref>`, you will probably get `encoded_value<std::string>`. This creates a problem since you don't want to define your codec using `encoded_value<std::string>` since that would have to create a copy when decoding. However, if you define your codec using `encoded_value<ref>` you instead run into the problem that you would have to re-validate your newly encoded `encoded_value<std::string>` when you create an `encoded_value<ref>` from it for putting into your struct that you then encode.

This could be solved by also including an implicit conversion from `encoded_value<std::string>` to `encoded_value<ref>`. Is there anything that would actually return `encoded_value<std::vector<std::uint8_t>>`?

**Further question 2:**
Does it make sense for `encoded_value<ref>` and `encoded_value<std::string>` to share a template, really? I could see both having quite different semantics leading us to have separate specializations for both anyway...

**Comment by punchfox** [comment]
It has to be default constructible for the example I used, to be put in the `unordered_map` and used with `[]`. But as you say, `"{}"` is probably a better default than `"null"`. When we move the string out of the `encoded_value`, if it leaves an empty string, we could catch that in the `data()` accessor and return a static `"{}"` string instead, to ensure it is valid JSON. We'd run into trouble if it left garbage in the source string, e.g.,

```c++
auto src = std::string("hello world");
const auto tgt = std::string(std::move(src));
src == "gorble garble";  // :-O
```

Regarding `std::vector`, we sometimes want to use the `encoded_value` with the body of a Cosmos request. But I'm actually thinking of making that into a `std::string` anyway.

**Comment by emil-e** [comment]
I'm not entirely sure I think a moved from `encoded_value` needs to be valid, though.

**Comment by punchfox** [comment]
What about the name of `raw_t`? I like `any_value_t`, coupled with renaming the current `any_t` codec to `any_codec_t`. Thoughts?

**Comment by pereckerdal** [comment]
**Overall feedback:** I don't really understand the benefits of this change, at least not when considering added complexity vs benefit. If it is only to enhance `raw_t`, it seems a bit overengineery to change the return type of `encode` to some new thing that people will have to learn to use the library. Are there other benefits to this than a faster validated `raw` serialization in the case that the serialized value has been encoded in the library somewhere else?

I also don't think I fully understand the point of the `type` enum class.

I think this change would have a smaller impact on the perceived complexity of the API if it didn't change the core `encode` functions.

**Regarding `any_codec_t`:** I think renaming `any_t` to `any_codec_t` would make sense, even if nothing else proposed here is done.

**Regarding if the object needs to be valid after move:** I agree with Emil: I think it's ok to leave garbage after a move, as long as the destructor doesn't do bad things. But then again, making it return "{}" after move isn't so bad is it?

**Regarding Question 2:** I think `"null"` makes more sense as a default than `"{}"`, but I don't have a strong opinion.

**Comment by emil-e** [comment]
To implement the behavior of returning `null` after a move, you would have to add a boolean to indicate the fact that the value has been move (is empty). Relying on a string being empty after moving is unlikely to be portable. Since small string optimization exists, it's not even particularly weird to not touch the moved from string if it fit in the internal buffer.

**Comment by punchfox** [comment]
Peck, having a type that we are certain will contain valid JSON allows us to skip that check when using the `raw_t`/`any_value_t` codec. We've had a bug already trying to use the `raw_t` codec with a string that was not valid JSON (provided by other parts of the code), so I definitely want to add that check.

I agree that `"null"` is probably a better default. The JSON spec is not super clear, and in any case, we use this data type to represent JSON values, which definitely include `null`.

You can disregard the `type` enum class for now. It does not affect the rest of the proposal and should really be a proposal of its own.

Let's define that the `encoded_value` should not be used after moving out of it, since the contents are undefined. This is similar to how most types work anyway when moving, e.g., you wouldn't use a `std::string` that has been moved, since it could contain anything from an empty string, the same string (short string optimisation), to complete garbage (maybe debug builds on some implementation).

I will implement and put up a PR for us to look at.

**Comment by punchfox** [comment]
This was implemented in #18, with some changes discussed offline, like not changing the return value of `json::encode`, instead introducing a new function `json::encode_value`; mainly because I ran into what seems to be a bug in Clang. Also keeps things a bit simpler for people really just wanting a `std::string` from the `encode` function.
